frontend/src/components/HeaderLoginInfo.jsx

```diff
@@ -0,0 +1,61 @@
+import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
+import {faUser} from "@fortawesome/free-solid-svg-icons";
+import StyledFirebaseAuth from "react-firebaseui/StyledFirebaseAuth";
+import firebase from "firebase";
+import {updateDatabase} from "../book/store";
+import Popup from "reactjs-popup";
+
+const HeaderLoginInfo = ({ email }) => {
+  return email ?
+    <><FontAwesomeIcon icon={faUser}/> {email}</> :
+    <Popup
+      trigger={
+      <button className="btn btn-primary">
+        <FontAwesomeIcon icon={faUser}/> Login / Sign up
+      </button>
+    }
+    modal
+    closeOnDocumentClick
+  >
+    {close =>
+      <StyledFirebaseAuth
+        uiConfig={{
+          signInOptions: [
+            {
+              provider: firebase.auth.EmailAuthProvider.PROVIDER_ID,
+              requireDisplayName: false,
+            },
+            // TODO not working because of cross origin isolation
+            // firebase.auth.GoogleAuthProvider.PROVIDER_ID,
+            // firebase.auth.FacebookAuthProvider.PROVIDER_ID,
+            // firebase.auth.GithubAuthProvider.PROVIDER_ID,
+          ],
+          autoUpgradeAnonymousUsers: true,
+          callbacks: {
+            // Avoid redirects after sign-in.
+            signInSuccessWithAuthResult: () => {
+              close();
+              window.location.reload();
```

Comment:
I feel like this is supposed to be possible without reloading the page, but I don't want you to waste a lot of time trying to figure out how. If we need to reload, just remove that comment.

Reply:
One issue with reloading is that `loadUser` is supposed to merge the local redux state with data from firebase, so logging in to an existing account preserves progress you made while anonymous. `loadUserAndPages` needs to complete and save the merged data back to the database before the page reloads and loses that state. That sort of seems to be happening but I can't see why and I'm worried it only seems to work locally where it doesn't need to wait for the internet.

Reply:
This component should update once a users' state changes, so I was confused as to why it wasn't happening (no updates were detected in redux) and added a refresh for this to happen. Upon further inspection, it seems that
```
firebase.auth().onAuthStateChanged(...)
```
in _src/book/store.js_ triggers when user state is changed, but only on sign in, sign out or token change (expiry or password change). I've looked into it and it seems that the best solution is to replace this method with
```
firebase.auth().onIdTokenChanged(...)
```
which takes the same parameters and triggers upon sign in, sign out and all token refresh events (meaning it'll trigger on sign ups too!). I've tried it and it seems to be working fine, but It'd be best if you looked into it too.

Reply:
It sounds like onIdTokenChanged also just gets called periodically when the token is refreshed without any user action. I don't like the sound of that.

From experimenting, it seems like signInSuccessWithAuthResult only gets called when signing up with a new account, not signing into an existing account (I assume that goes through signInFailure). And in that case onAuthStateChanged never gets called. So what if we explicitly made onAuthStateChanged and signInSuccessWithAuthResult both call the same function (which in turn calls loadUser)?

Reply:
Reverted to onAuthStateChanged and explicitly updated store with user data on sign in success.